### ISSUE TITLE
Support for vLLM running in OpenAI compatible mode

### ISSUE THREAD

**Issue by vecorro** [triton-inference-server/server#6583]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to run vLLM emulating the OpenAI compatible API to use vLLM as a drop-in replacement of ChatGPT.

**Describe the solution you'd like**
I'd like Triton allow me run vLLM as indicated in [vLLM documentation](https://vllm.readthedocs.io/en/stable/getting_started/quickstart.html#openai-compatible-server)

Example:

`python -m vllm.entrypoints.openai.api_server --model facebook/opt-125m`

**Describe alternatives you've considered**
It is possible to use the REST API, however, for developers already leveraging OpenAI and serving open-source LLMs using the OpenAI API would allow a faster path for replacement of OpenAI

**Comment by nnshah1** [comment]
@vecorro It's a bit challenging as the OpenAI API is a moving target and already the completion api is legacy and out of date with many applications. As triton as an inference server serves many types of inference we've focused on providing ways to customize endpoints also with the expectation that many actual deployments would use an additional api gateway to translate from a service api to the inference api (for example langchain).

that being said - your perspective is important. Is the legacy completion api enough or would we need to support the newer apis as well? Do you think integration with a project like langchain or other llm projects would be a viable approach or would triton need to provide the interface directly?

**Comment by chymian** [comment]
@nnshah1 
> that being said - your perspective is important. Is the legacy completion api enough or would we need to support the newer apis as well? Do you think integration with a project like langchain or other llm projects would be a viable approach or would triton need to provide the interface directly?

The key here, to be usefull to the community, is actuality & compatibility. every other often used tool, like LiteLLM (middlware/API-Proxy) or loaders, like ooba's TGI, vLLM, FastChat are on the run to implement `functions` and to pick up with v1, since this is essential for local LLMs to be a replacement for the costly openai-stuff.
There are a few very popular projects which use the very latest features heavly, like Mircosofts Autogen and MemGPT.

IMHO: I have been testing all these and just had a peek into triton, but already assume, it would be the much supirior solution of running multiple LLMs locally - IF there would be an out of the box compatibility with the openAI-API.
Maybe teaming up with https://litellm.ai/ is a quick solution.

**Comment by ishaan-jaff** [comment]
@chymian @nnshah1 im the litellm maintainer - what do you need from us ?

**Comment by vecorro** [comment]
I agree with @nnshah1; the key is actuality and compatibility, and I would add ease of use. For experimentation/ded purposes, vLLM is excellent as it allows you to try an LLM without the complexity and time required by TensorRT-LLM. **Making the existing vLLM implementation of the legacy OpenAI API can be good enough**; meanwhile, vLLM evolves to support newer versions of the OpenAI API. I'm on the VMware team working with NVIDIA on the private AI initiative. From conversations with customers looking for ways to run LLMs on-premises, they are getting started on that journey, so a quick drop-in replacement of OpenAI's models would facilitate things for them. Thanks all!

**Comment by nnshah1** [comment]
@npuichigo has mentioned an integration that could be useful here:

https://github.com/NVIDIA/TensorRT-LLM/discussions/591

**Comment by npuichigo** [comment]
> @npuichigo has mentioned an integration that could be useful here:
> 
> [NVIDIA/TensorRT-LLM#591](https://github.com/NVIDIA/TensorRT-LLM/discussions/591)

https://github.com/npuichigo/openai_trtllm provides an OpenAI-like API for trtllm triton backend, but I think vllm in triton would be something alike.

**Comment by dyastremsky** [comment]
Closing issue due to inactivity. Please reopen if you would like to follow up with this issue.

**Comment by BodhiHu** [comment]
Hello,

Seems it's supporting vllm now:
https://github.com/triton-inference-server/tutorials/blob/main/Quick_Deploy/vLLM/README.md#deploying-a-vllm-model-in-triton

But can we use the vllm OpenAI APIs ?

Thanks a lot

**Comment by nnshah1** [comment]
We don't currently support it directly - but are still thinking through the best ways to add a compatible API in a way that we can maintain.

**Comment by BodhiHu** [comment]
Thank you for the clearance~~
though I think we could simply adapte the vLLM `vllm.entrypoints.openai.api_server` to the [triton http endpoint](https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_generate.md#httprest).

**Comment by panpan0000** [comment]
is there any workaround so far ?

> Thank you for the clearance~~ though I think we could simply adapte the vLLM `vllm.entrypoints.openai.api_server` to the [triton http endpoint](https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_generate.md#httprest).

**Comment by nnshah1** [comment]
> @chymian @nnshah1 im the litellm maintainer - what do you need from us ?

Is there a guide on adding an llm provider to litellm or a proscribed starting point / skeleton?

**Comment by anubhav-agrawal-mu-sigma** [comment]
Is this planned for any specific release?

**Comment by nnshah1** [comment]
@anubhav-agrawal-mu-sigma - we are currently planning a tutorial showcasing how to create an open ai compatible triton server using triton's in-process python api - that tutorial is planned for September time frame.

**Comment by catle2aurecon** [comment]
@nnshah1 : I am interested in the mentioned tutorial, please let me know when you have it online. Many thanks !

**Comment by agladboy** [comment]
@nnshah1 I’m excited to see the tutorial! Would you please to let me know when it will be available? Thanks in advance!

**Comment by BrandesDenis** [comment]
@nnshah1 I'm interested too!

**Comment by nnshah1** [comment]
You can follow along the current preview PR here:

https://github.com/triton-inference-server/server/pull/7561 

Some things may change- so please consider this as BETA until we finalize a few of the internal structure - but this will be the basis.

**Comment by mohamad-tohidi** [comment]
Hello
is this feature implemented in PyTriton as well?


if it is i think you shoul update this document, to include an example

https://github.com/triton-inference-server/tutorials/blob/main/Quick_Deploy/vLLM/README.md#deploying-a-vllm-model-in-triton
